### ISSUE TITLE
Shorter allele not removed

### ISSUE THREAD

**Issue by schorlton** [B-UMMI/chewBBACA#52]
Created scheme with the attached genomes 
[creation_genomes.tar.gz](https://github.com/B-UMMI/chewBBACA/files/4006772/creation_genomes.tar.gz)
and training file.
[NCTC11397.trn.gz](https://github.com/B-UMMI/chewBBACA/files/4006776/NCTC11397.trn.gz)

Discovered that for locus
[GCA-900312965-protein756.fasta.zip](https://github.com/B-UMMI/chewBBACA/files/4006771/GCA-900312965-protein756.fasta.zip), alleles 2 and 20 are identical, except that allele 20 is 18bp shorter than allele 2. [0DX541TX11N-Alignment.txt](https://github.com/B-UMMI/chewBBACA/files/4006768/0DX541TX11N-Alignment.txt)
I thought that allele 20 should have been removed?

**Comment by ramirma** [comment]
Thank you for using chewBBACA @schorlton. The behavior you describe is exactly the expected behavior of chewBBACA. The two alleles are not "identical" because allele 20 is 18bp shorter than allele 2. Allele 20 is contained in allele 2 but for chewBBACA these are two different alleles. This happened probably because it was not possible to find an open reading frame that extended the allele to the size of allele 2. In your case an alternate start codon was found 18bp downstream of the start codon used in allele 2 possibly due to a change in those 18bp that generated a stop codon.
Remember that with the default parameters chewBBACA will allow alleles to vary in length +/-20% from the mode of the allele size. This is clearly seen in the file you sent which has alleles of variable sizes.
I hope this clarifies your question.

**Comment by schorlton** [comment]
Thanks for your help! Maybe you could help me understand what this line refers to from the wiki, as it seems to refer to this situation?
"The first step in each pairwise comparison is the removal of CDSs that are contained within larger CDSs, which occurs if the CDSs are identical but use different start codons."
https://github.com/B-UMMI/chewBBACA/wiki/1.-Schema-Creation

Something similar is also in your publication. As the CDSs are only clustered into scheme loci after removal of shorter CDSs (as per the publication), I did not expect shorter, exact alleles within each scheme locus.

**Comment by ramirma** [comment]
These lines refer to the schema creation part and not the allele calling part. This approach will guarantee that we keep the largest alleles when creating a schema, and not their smaller, possibly truncated versions. When performing allele calling we no longer impose this limitation.

**Comment by schorlton** [comment]
Ah I think I understand: for each genome, because Prodigal will call a CDS within another CDS, this shorter CDS is removed. If, in another genome, only the shorter CDS is present, it will be kept as an allele of the same locus from the other genome. Thanks again for your help!

**Comment by schorlton** [comment]
Alright sorry to keep going with this, but I still don't completely understand. We can go to gitter if you prefer.

The following 2 assemblies had different alleles called (results: [cgMLST.tsv.gz](https://github.com/B-UMMI/chewBBACA/files/4007281/cgMLST.tsv.gz) or pre-extraction result: [results_alleles.tsv.gz](https://github.com/B-UMMI/chewBBACA/files/4007293/results_alleles.tsv.gz)); ([germanA.fa.gz](https://github.com/B-UMMI/chewBBACA/files/4007283/germanA.fa.gz) had allele 20, [germanB.fa.gz](https://github.com/B-UMMI/chewBBACA/files/4007284/germanB.fa.gz) had allele 2) at GCA-900312965-protein756.fasta (Allele fasta posted above), but they both contain the full length 570bp of allele 2. Blast results: 
[GermanB_2.txt](https://github.com/B-UMMI/chewBBACA/files/4007288/GermanB_2.txt)
[GermanB_20.txt](https://github.com/B-UMMI/chewBBACA/files/4007289/GermanB_20.txt)
[GermanA_2.txt](https://github.com/B-UMMI/chewBBACA/files/4007290/GermanA_2.txt)
[GermanA_20.txt](https://github.com/B-UMMI/chewBBACA/files/4007291/GermanA_20.txt)

To call alleles, I used chewBBACA 2.1.0 with the following command:
 chewBBACA.py AlleleCall -i genomes/ -g ~cgMLSTschema_withlocation.txt -o cgMLST_typing --cpu 30 --ptf ~/Data/dip/cgMLST/prodigal_training/NCTC11397.trn
where the scheme is the location of each of the 1474 fastas previously defined as cgMLST loci.

[RepeatedLoci.txt](https://github.com/B-UMMI/chewBBACA/files/4007294/RepeatedLoci.txt)
[logging_info.txt](https://github.com/B-UMMI/chewBBACA/files/4007295/logging_info.txt)

**Comment by ramirma** [comment]
No problem @schorlton. In my initial response I did not realize that you were talking about the schema creation part and not the allele calling part. Looking more carefully at the files you sent, I would also like to point out that the locus in question is on the low range (570nt for allele 2) and that the 18nt difference is a 3% variation. This combination of relatively small size and significant size variation may have caused a lower BSR, justifying the inclusion of the two alleles as "representative" alleles.

A separate issue is why Prodigal is calling the CDS differently in each assembly and one I am afraid I will not be able to help you with. As you know "Prodigal is an unsupervised machine learning algorithm" and I suspect it will be difficult to pinpoint why this happened. In chewBBACA we decided to have the Prodigal predictions as a base for allele identifications and this may have generated this particular issue.
